src/rgw/driver/sfs/sqlite/dbconn.h

```diff
@@ -254,13 +253,16 @@ inline auto _make_storage(const std::string& path) {
 }
 
 using Storage = decltype(_make_storage(""));
+using StorageRef = Storage*;
 
 class DBConn {
  private:
-  Storage storage;
+  std::unordered_map<std::thread::id, Storage> storage_pool;
+  std::thread::id main_thread;
+  mutable std::shared_mutex storage_pool_mutex;
 
  public:
-  sqlite3* first_sqlite_conn;
+  std::unordered_map<std::thread::id, sqlite3*> all_sqlite_conns;
```

Comment:
Can we make this private? It has a non trivial access story (pull mutex?)

Do we actually need the thread id here? I can't find any references. Maybe a vector is enough. Still needs synchronized access though.

Reply:
Second question first, because it's the easy one :-) The thread id is there because `get_sqlite_conn()` uses it to get the `sqlite3*` specifically associated with the calling thread. Same mechanism as `get_storage()`, just a different map. I found keeping two maps made some of the code easier to read rather than one map with a `std::pair<Storage, sqlite*>` or a struct, but either approach would work.

As to can we make it private... The current code inside DBConn actually only uses the main thread's `sqlite*` for a couple of things during startup to do with updating metadata. The other use of `all_sqlite_conns` is in `SFSStatusPage::render`, where we use it to get db status for all threads. Having it public made that really easy and lightweight, but you're right, it would then need synchronized access, dammit.

I guess the safest thing to do under the circumstances is make that map private (or fold it into the `storage_pool` map), and add a method that returns a vector created as necessary from the contents of the map.  Something like this (untested):
```
std::vector<sqlite*> all_sqlite_conns() const {
  std::vector<sqlite*> conns;
  std::shared_lock lock(storage_pool_mutex);
  for (auto p : all_sqlite_conns) {
    conns.emplace(p.second)
  }
  return conns;
}
```

What do you think?

Reply:
I went with the alternative :-)